ToolAnnotations

bhosmer-ant opened the pull request:
<!-- Provide a brief summary of your changes -->

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->

`ToolAnnotations` provides additional optional properties for describing a tool's behavior beyond its `inputSchema` and `description` (the latter being intended for model hinting, rather than user presentation). 

`ToolAnnotations` contains both operational and display properties. Display properties are defined in the `DisplayAnnotations` interface, which `ToolAnnotations` inherits from. (This anticipates the use of `DisplayAnnotations` elsewhere, as proposed in #201 .)

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
N/A

## Breaking Changes
<!-- Will users need to update their code or configurations? -->
None

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [ ] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

PederHP commented on schema/draft/schema.json:
```diff
@@ -2065,6 +2069,29 @@
             "type": "object"
         },
+        "ToolAnnotations": {
+            "description": "A tool annotation categorizes the tool and describes its operational behavior.\n\nNote that this information is intended for use by client infrastructure,\nand is distinct from what a model will extract/infer from the tool's description\nproperty.",
```
Maybe use the term 'application' instead of 'client infrastructure' to align with the concepts used elsewhere in the spec?

PederHP commented:
Very happy to see Tool annotations introduced in a generic fashion.

PederHP commented on schema/draft/schema.ts:
```diff
@@ -677,6 +677,47 @@ export interface ToolListChangedNotification extends Notification {
+/**
+ * A tool annotation categorizes the tool and describes its operational behavior.
+ * 
+ * Note that this information is intended for use by client infrastructure,
```
Maybe use the term 'application' instead of 'client infrastructure' to align with the concepts used elsewhere in the spec?

I would also suspect that many host applications will choose to show these, e.g. Category directly to users. Not sure that's worth mentioning in the schema though, but it would be valuable to comment on elsewhere in docs.

bhosmer-ant replied:
Good point, it's worth mentioning applications here.

bhosmer-ant replied:
(Comment updated to avoid over-specifying usage)

PederHP commented on schema/draft/schema.ts:
```diff
@@ -677,6 +677,47 @@ export interface ToolListChangedNotification extends Notification {
+ * A tool may specify a category outside the predefined set.
+ */
+export type ToolCategory =
+  "search" |
+  "repl" |
```
Why these two particular categories as predefined? And if they are especially important should it not be document what the expected meaning of the category is?

bhosmer-ant replied:
Yeah good questions. Re "why these two", they were just here to seed the enum/prompt discussion, and after some further thought/discussion I don't think the notion of "category" as a tool annotation really works:
* MCP shouldn't be in the business of defining a canonical tool ontology, which is what this would have to amount to in order to be concretely useful. (This is to your point - without further specification these terms are too ambiguous, but designing a catalog of canonical, future-proof definitions for high-level concepts like "search tool" is fraught in a bunch of ways.)
* there's also a granularity problem: often a tool will be part of a group of tools that together support a particular use-case like "search" or "repl" or "database". Here again a flat enum of category descriptors (no matter how carefully designed) isn't really expressive enough to describe the tool properly. 
About to push a revision without `category` - thanks for the feedback!

bhosmer-ant replied:
Update: removed `category`, per the shortcomings described above: TLDR it's fatally underpowered as a classifier (not descriptive enough, and even with supplemental definitions/documentation, doomed to chasing an evolving world of tools).

olaservo commented on schema/draft/schema.json:
```diff
@@ -2065,6 +2069,29 @@
+                    "type": "string"
+                },
+                "extraInfo": {
+                    "additionalProperties": {},
```
I might have missed a discussion but I'm a little unclear on the value of having generic fields like `extraInfo`.  

Also I've noticed the term `metadata` used for other open-ended fields (like in sampling requests), so if there does need to be something like this added to tool annotations, would it make sense to use that generic term instead?

bhosmer-ant replied:
You're right, the intended use was underspecified. The original motivation began with the suspicion that `category` would often need refinement/qualification to be useful as a classifier, and became generic from there. But it doesn't really solve the basic issue with `category` (see comment above), so I've removed both. (For arbitrary extra information, the protocol actually allows for any object to be extended ad-hoc.)

Also though you're right, `metadata` would have been a more idiomatic name had it stayed :)

jspahrsummers reviewed (changes requested):
Sorry, I have some worries here, particularly around the trust model. I do think adding more expressivity here will be a good thing, but we need to be careful not to promise something that will actually be broken in practice.

bhosmer-ant commented on schema/draft/schema.ts:
```diff
@@ -696,6 +696,48 @@ export interface ToolListChangedNotification extends Notification {
+   */
+  title?: string;
+
+  icon?: {
+    data: string; // base64-encoded image data
+    contentType: string; // MIME type of the image
+  };
+
+  /**
+   * A URL from which a client can fetch an icon to represent this object.
+   */
+  icon_url?: string;
```
Makes sense - I don't really have a strong opinion, though I could imagine genuinely different tools wanting their own icons and tool bundles that work together on the same task just shipping the same icon. 

Other related things:
- @jerome3o-anthropic mentioned tool icons as potentially useful, not sure if he has a strong opinion though.
- #201 proposes adding icons to `Annotated`. I guess the use case there is for more ephemeral objects, so not directly analogous w.r.t. the  "too noisy" problem, but fyi in case you wanted to chime in there.

I'll default to pulling them out of here absent followup discussion.

jspahrsummers replied:
Yeah, I do lean toward custom icons being acceptable for _data_ types, but less so for _tools_ (which IMO should bear the branding of their server).

bhosmer-ant commented on schema/draft/schema.ts:
```diff
@@ -696,6 +696,48 @@ export interface ToolListChangedNotification extends Notification {
+ * Annotations that provide display-facing and operational information for a Tool.
+ */
+export interface ToolAnnotations extends DisplayAnnotations {
+  /**
+   * If true, this tool does not perform writes or updates,
+   * or otherwise change server-side state in ways that would
+   * be visible in subsequent tool calls.
+   *
+   * If not set, this is assumed to be false.
+   */
+  readOnly?: boolean;
```
First, totally agreed on emphasizing trust - I'll add language about this to every property. (Also just realized I haven't documented this at all, will add.)

Re contingent properties like "destructive" and "idempotent" (and I agree we should add the obvious ones eagerly) - it's that classic data modeling situation where we could either use a variant type (pro: no nonsense combinations) or a product of bools (pro: only primitive values). 

I guess the fancy version in TS would be something like
```
type ReadOnly = {
  readonly: true;
};

type Writable = {
  readonly: false;
  destructive: boolean;
  idempotent: boolean;
};

type OperationMode = ReadOnly | Writable;
```
Personally I'd be inclined to do this kind of thing rather than just a plank of bools, but interested in your thoughts.

bhosmer-ant replied:
@PederHP thanks for calling this out - added

PederHP commented on schema/draft/schema.json:
```diff
@@ -1980,6 +2013,43 @@
             ],
             "type": "object"
         },
+        "ToolAnnotations": {
```
Having some kind of additional properties mechanism to have free-form annotations would be very useful. The use case is having internal MCP servers that we use with internal MCP clients (that are back-end services). The annotations we'd use (which could include rate-limiting information, information sensitivity classifiers, etc.) would likely not be general enough to ever find their way into the spec, but they'd of great value to us.

The alternative is that we need maintain this information elsewhere - or do a branch of the SDK(s) and add the fields as what we do is all internal, but that's not optimal either. Free-form annotations are very useful for MCP-based infrastructures.

jspahrsummers replied:
All objects in the MCP spec are intentionally open-ended, so it's always possible to add custom fields that only your server and client implementations understand. (No forking required!)

hackjammer replied:
Another drive by comment, I just stumbled on this PR whilst considering a parallel internal implementation. Some thoughts:
* Information sensitivity classifiers is one of the examples we had. So this is always going to need to be extensible for different businesses with different definitions and risks
* Looks like effectHints is not extensible as currently implemented? We had other property ideas e.g. "appendOnly" or "irrevocable". If we in general want this to be extensible then maybe effectHints should be too
* In general it might be good to give examples of how this should be extended for custom metadata / hints and guidance on when to share back standards

PederHP replied:
> All objects in the MCP spec are intentionally open-ended, so it's always possible to add custom fields that only your server and client implementations understand. (No forking required!)

I hope you don't mind a comment on this on a merged PR, but it seemed like the obvious place: this is true for weakly-typed languages, like Python or TypeScript, but for C# it makes a difference if the schemas explicitly has metadata and/or kwargs-style additional properties (as some types do, but not all). 

So I think either adding a basetype that makes this explicit, or adding it explicitly to types where it is likely to be used (like Tools) would be very helpful. Especially for cases where a server is Python/TypeScript and the Client is C#. 

I assume similar concerns are valid for Rust and Java. Perhaps I should open an issue about consistency in terms of how open-ended types are handled in the scehma?

jspahrsummers commented on schema/draft/schema.ts:
```diff
@@ -696,6 +696,48 @@ export interface ToolListChangedNotification extends Notification {
+   *
+   * If not set, this is assumed to be true.
+   */
+  openWorld?: boolean;
```
I think the information itself, _if it could be trusted_, would be very useful, but I wonder how a client makes use of this flag knowing that it's _not_ trustable.

I guess that maybe degrades to, "Clients should apply a high scrutiny bar to all untrusted servers and ignore most of these annotations, until the user tells them not to."

bhosmer-ant replied:
100%, actually I think "Clients should ignore annotations from untrusted servers" applies to **all** annotations, even `title` - but especially the ones that describe operational properties. Worth strengthening the language. 

Separately, do we discuss the question of server trust elsewhere in the spec? (Not as a matter of definition, but operationally.)

jspahrsummers replied:
There are some general mentions—e.g., https://spec.modelcontextprotocol.io/specification/2024-11-05/#security-and-trust--safety

jspahrsummers commented on schema/draft/schema.ts:
```diff
@@ -696,6 +696,50 @@ export interface ToolListChangedNotification extends Notification {
+   * If not set, this property is assumed to have the value:
+   *  { readOnly: false, destructive: true, idempotent: false }
+   */
+  effectHints?:
+    | {
+        /* Tool is read-only. */
+        readOnly: true;
+      }
+    | {
+        /* Tool is read-write. */
+        readOnly: false;
+        /* If true, tool may perform destructive updates to its environment, as opposed to only additive updates. Default: true */
+        destructive?: boolean;
+        /* If true, repeated calls with the same arguments will have no additional effects. Default: false */
+        idempotent?: boolean;
+      };
```
Nit: I think naming these types is helpful, for documentation _but also_ to create more distinct entities in the JSON Schema, and possibly assist codegen.

bhosmer-ant replied:
Question, are subobjects also open-ended in the spec? If not, the extensibility discussion above is actually shifting my preference back towards simple bool properties (with appropriate documentation of when variant properties should be ignored) for the obvious reason...

jspahrsummers reviewed (dismissed):
Thanks, LGTM!

bhosmer-ant commented:
@jspahrsummers looks like my post-approve changes caused your review to be dismissed, sorry about that, can I get you to TAL? Here's what's changed:
- switched from the variant-type setup to simple boolean properties. The data model is less precise in a closed world but I think the discussion here shines a light on the relative benefit of closed-world precision vs ease of extensibility/coherence of custom properties.
- added trusting-annotations callout to the Security and Trust & Safety section in _index.md
- added `annotations` to the property list in TOOLS.md (and another callout)

jspahrsummers reviewed (dismissed):
SGTM

bhosmer-ant commented on docs/specification/draft/server/tools.md:
```diff
@@ -181,6 +181,10 @@ A tool definition includes:
+- `annotations`: optional properties describing tool behavior
+
+{{< callout type="warning" >}} For trust & safety and security, clients **MUST** consider
+tool annotations to be untrusted unless they come from trusted servers.
```
argh yes confirmed and fixed
